### PR TITLE
Dataproc Provisioner with Autoscale (Experimental Feature)

### PR DESCRIPTION
**Pull request by shiva-cirus**
Provisioner with Autoscale option.

### CODE REVIEW

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/common/DataprocUtils.java [hunk 1]
```diff
@@ -22,6 +22,8 @@
 import com.google.cloud.storage.StorageBatch;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+import java.util.HashMap;
```
Comment (chtyim) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/common/DataprocUtils.java:26]: The current style guide has a empty line between non-java and java imports. You can refer to existing classes.

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/common/DataprocUtils.java [hunk 2]
```diff
@@ -58,6 +60,25 @@ public static void deleteGCSPath(Storage storageClient, String bucket, String pa
...
+   * @param configValue
+   * @param delimiter
+   * @param keyValueDelimiter
+   * @return
+   */
+  public static Map<String, String> parseKeyValueConfig(String configValue, String delimiter,
+                                                        String keyValueDelimiter) {
+    Map<String, String> map = new HashMap<>();
+    for (String property : configValue.split(delimiter)) {
+      String[] parts = property.split(keyValueDelimiter);
```
Comment (chtyim) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/common/DataprocUtils.java]: Call `split(keyValueDelimiter, 2)`.

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/common/DataprocUtils.java [hunk 3]
```diff
@@ -58,6 +60,25 @@ public static void deleteGCSPath(Storage storageClient, String bucket, String pa
...
+      String key = parts[0];
```
Comment (chtyim) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/common/DataprocUtils.java:83]: Validate the array size is 2 before using it.

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocConf.java [hunk 4]
```diff
@@ -243,6 +261,40 @@ SSHPublicKey getPublicKey() {
...
+      return Collections.emptyMap();
+    }
+    return DataprocUtils.parseKeyValueConfig(clustermetadata, ";", "\\|");
+  }
+
+  List<String> getNetworkTags() {
+    if (Strings.isNullOrEmpty(networktags)) {
+      return Collections.emptyList();
+    }
+    return Arrays.stream(Objects.requireNonNull(networktags).split(","))
```
Comment (chtyim) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocConf.java]: No need to have requireNonNull check

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocConf.java [hunk 5]
```diff
@@ -243,6 +261,40 @@ SSHPublicKey getPublicKey() {
...
+      .map(String::trim)
+      .collect(Collectors.toList());
+  }
+
+  List<String> getInitActions() {
+    if (Strings.isNullOrEmpty(initactions)) {
+      return Collections.emptyList();
+    }
+    return Arrays.stream(Objects.requireNonNull(initactions).split(","))
```
Comment (chtyim) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocConf.java]: No need to have requireNonNull check

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocConf.java [hunk 6]
```diff
@@ -390,12 +442,22 @@ static DataprocConf create(Map<String, String> properties, @Nullable SSHPublicKe
     String gcpCmekKeyName = getString(properties, "encryptionKeyName");
     String gcpCmekBucket = getString(properties, "gcsBucket");
 
+    String clustermetadata = getString(properties, "clustermetadata");
```
Comment (chtyim) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocConf.java]: Should be camel case `clusterMetadata`. Same for the string value in the property

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocConf.java [hunk 7]
```diff
@@ -390,12 +442,22 @@ static DataprocConf create(Map<String, String> properties, @Nullable SSHPublicKe
...
+    String networktags = getString(properties, "networktags");
```
Comment (chtyim) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocConf.java]: Camel case

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocConf.java [hunk 8]
```diff
@@ -390,12 +442,22 @@ static DataprocConf create(Map<String, String> properties, @Nullable SSHPublicKe
...
+    String initactions = getString(properties, "initactions");
```
Comment (chtyim) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocConf.java]: Camel case

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocConf.java [hunk 9]
```diff
@@ -390,12 +442,22 @@ static DataprocConf create(Map<String, String> properties, @Nullable SSHPublicKe
...
+    boolean enableautoscaling = Boolean.parseBoolean(properties.getOrDefault("enableautoscaling",
```
Comment (chtyim) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocConf.java]: Just call it `autoScaling`

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocConf.java [hunk 10]
```diff
@@ -88,14 +93,21 @@
   private final SSHPublicKey publicKey;
   private final Map<String, String> dataprocProperties;
 
+  private final String clustermetadata;
+  private final String networktags;
+  private final String initactions;
+  private final boolean enableautoscaling;
+  private final String autoscalingpolicy;
```
Comment (chtyim) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocConf.java]: Why need both enableAutoScaling and autoScalingPolicy?? We can just have the `autoScalingPolicy`. If it is null, then there is no auto-scaling.

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocProvisioner.java [hunk 11]
```diff
@@ -155,6 +156,27 @@ public void validateProperties(Map<String, String> properties) {
...
+    if (networkTags.size() > 64) {
+      throw new IllegalArgumentException("Exceed Max number of tags. Only Max of 64 allowed. ");
+    }
+    //Validate if AutoScaling is enabled and Policy is provided
+    if (conf.isAutoPolicyEnabled()) {
+      if (conf.getAutoScalingPolicy() == null || conf.getAutoScalingPolicy().isEmpty()) {
+        throw new IllegalArgumentException("Autoscaling is enabled but no AutoScaling policy provided. " +
+                                             "Provide AutoScaling Policy.");
+      }
+
```
Comment (chtyim) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocProvisioner.java]: Remove extra new line

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/common/DataprocUtils.java (hunk 1 above)
Comment (shiva-cirus) [reply cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/common/DataprocUtils.java:26]: Done

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/common/DataprocUtils.java (hunk 2 above)
Comment (shiva-cirus) [reply cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/common/DataprocUtils.java]: Done

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/common/DataprocUtils.java (hunk 3 above)
Comment (shiva-cirus) [reply cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/common/DataprocUtils.java:83]: Done

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocConf.java (hunk 6 above)
Comment (shiva-cirus) [reply cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocConf.java]: Done

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocConf.java (hunk 7 above)
Comment (shiva-cirus) [reply cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocConf.java]: Done

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocConf.java (hunk 8 above)
Comment (shiva-cirus) [reply cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocConf.java]: Done

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocConf.java (hunk 10 above)
Comment (shiva-cirus) [reply cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocConf.java]: This was implemented as per Spec from Bhooshan. Need to check with him if we have to remove it.

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocProvisioner.java (hunk 11 above)
Comment (shiva-cirus) [reply cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocProvisioner.java]: Done

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocConf.java (hunk 5 above)
Comment (shiva-cirus) [reply cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocConf.java]: Done

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocConf.java (hunk 9 above)
Comment (shiva-cirus) [reply cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocConf.java]: Done

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/common/DataprocUtils.java [hunk 12]
```diff
@@ -58,6 +61,28 @@ public static void deleteGCSPath(Storage storageClient, String bucket, String pa
...
+   * @param delimiter
+   * @param keyValueDelimiter
+   * @return
+   */
+  public static Map<String, String> parseKeyValueConfig(String configValue, String delimiter,
+                                                        String keyValueDelimiter) {
+    Map<String, String> map = new HashMap<>();
+    for (String property : configValue.split(delimiter)) {
+      String[] parts = property.split(keyValueDelimiter, 2);
+      if (parts.length == 2) {
```
Comment (chtyim) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/common/DataprocUtils.java]: When length is not two, should throws an IllegalArgumentException with appropriate message tell the caller what's wrong and how to correct it.

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocConf.java [hunk 13]
```diff
@@ -390,12 +441,22 @@ static DataprocConf create(Map<String, String> properties, @Nullable SSHPublicKe
     String gcpCmekKeyName = getString(properties, "encryptionKeyName");
     String gcpCmekBucket = getString(properties, "gcsBucket");
 
+    String clusterMetaData = getString(properties, "clusterMetaData");
+    String networkTags = getString(properties, "networkTags");
+    String initActions = getString(properties, "initActions");
+    boolean autoScaling = Boolean.parseBoolean(properties.getOrDefault("autoScaling",
+                                                                             "false"));
```
Comment (chtyim) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocConf.java]: You don't need `getOrDefault`. Just call `get`. The `Boolean.parseBoolean` will return false if the given value is `null`.

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/common/DataprocUtils.java (hunk 12 above)
Comment (shiva-cirus) [reply cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/common/DataprocUtils.java]: Done

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocConf.java (hunk 13 above)
Comment (shiva-cirus) [reply cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocConf.java]: Done

**Pull request closed without merging**